### PR TITLE
Modified UserProject Model to handle invitations - #272

### LINKED ISSUE #272: [Backend] Modify UserProject model
We need to add some properties to UserProject Model to be possible work with invitation in the projects.

``` C#
    public User User { get; private set; } = new User(); // Already exists, change to optional
    public string Email { get; set; } // when user not exists
    public Project Project { get; private set; } = new Project();  // Already exists
    public Role Role { get; private set; }  // Already exists
    public Guid Token { get; set; }
    public bool Accepted { get; set; }
    public DateTime SentAt { get; set; }
    public DateTime? AcceptedAt { get; set; }
    public bool Expired { get; set; }
    public DateTime ExpiryDate { get; set; }
```
- [ ] create a index on token property

### PR DESCRIPTION
**Pull request by saksham-malhotra-27**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization

## Description
Added some properties to _UserProject_ Model to be possible work with invitation in the projects.

- Related Issue #272
- Closes #272

### CODE REVIEW

File: EasyFinance.Domain/AccessControl/UserProject.cs [hunk 1]
```diff
@@ -9,17 +9,29 @@ public class UserProject : BaseEntity
...
+        public string? Email { get; set; }      
         public Project Project { get; private set; } = new Project();
         public Role Role { get; private set; }
 
+        public Guid Token { get; set; }     
+        public bool Accepted { get; set; }        
+        public DateTime SentAt { get; set; }       
+        public DateTime? AcceptedAt { get; set; }
+        public bool Expired { get; set; } 
+        public DateTime ExpiryDate { get; set; }
```
Comment (FelipePSoares) [inline EasyFinance.Domain/AccessControl/UserProject.cs:33]:
```suggestion
        public Guid Token { get; private set; }     
        public bool Accepted { get; private set; }        
        public DateTime SentAt { get; private set; }       
        public DateTime? AcceptedAt { get; private set; }
        public bool Expired { get; private set; } 
        public DateTime ExpiryDate { get; private set; } 
```

All the models in this project has private set and you need to create a method to add the value and in this method before add the new value you will validate if the value is valid. The objetive here is use strong domains that keep valid eachself. 

You can also add in contructor each value to be added in the creation of the model calling this internal method to validate the values.

If you need some example you can look any model in this project.

**Comment by saksham-malhotra-27** [comment]
I think I understand why this is happening now. It's because I changed the logic to throw an error when both the user and email are null.

File: EasyFinance.Persistence/Mapping/AccessControl/UserProjectConfiguration.cs [hunk 2]
```diff
@@ -15,6 +15,12 @@ public override void ConfigureEntity(EntityTypeBuilder<UserProject> builder)
             builder.HasOne(p => p.User)
                 .WithMany()
                 .IsRequired();
```
Comment (FelipePSoares) [inline EasyFinance.Persistence/Mapping/AccessControl/UserProjectConfiguration.cs:17]:
```suggestion
                .WithMany();
```

I suggest you to remove this `isRequired` tag from user thinking that now the user can be null.

**Review by FelipePSoares — CHANGES_REQUESTED** [review]
Nice work, 

Just some suggestions.

**Comment by FelipePSoares** [comment]
Hey @saksham-malhotra-27!

Let me know if you have some doubt.

**Comment by saksham-malhotra-27** [comment]
@FelipePSoares little busy today

**Comment by FelipePSoares** [comment]
@saksham-malhotra-27 no worries. Take your time.

**Comment by saksham-malhotra-27** [comment]
Hey @FelipePSoares ! The error occurred because I was throwing an error when both the user and email were null in the constructor. So, in the tests where the user is null and there is no email, we'll receive that error instead of the predefined ones.

**Comment by FelipePSoares** [comment]
Yeah, probably yes, you are free to change the tests now that the behavior in the application changes no problem.

And for the validation message, please use the validationMessages inside the infrastructure project.

**Comment by saksham-malhotra-27** [comment]
For now, I have removed the error that was being thrown when both user and email were null and now nothing is giving an error in build. However, I would prefer if you could handle and update the validations and backend tests for this, as I'm a bit confused by .NET and may not be at my best due to feeling sick at the moment.
@FelipePSoares  thank you! Perhaps you can merge this one if the code looks right and then create a new issue for writing tests for this.

**Comment by FelipePSoares** [comment]
@saksham-malhotra-27 ,

Don't worry I'll finish this for you. Thank you very much for all your help with this task.

Get well soon.

**Review by FelipePSoares — APPROVED** [review]

**Comment by FelipePSoares** [comment]
Hey @saksham-malhotra-27 !

I finish it today. Thank you for your help again. 

Have a nice week.

**Comment by saksham-malhotra-27** [comment]
No problem @FelipePSoares and thank you for completing it !

**Pull request merged**